internal/controllers/topology/cluster/reconcile_state.go

```diff
@@ -438,6 +440,24 @@ func (r *Reconciler) reconcileCluster(ctx context.Context, s *scope.Scope) error
 		return errors.Wrapf(err, "failed to patch %s", tlog.KObj{Obj: s.Current.Cluster})
 	}
 	r.recorder.Eventf(s.Current.Cluster, corev1.EventTypeNormal, updateEventReason, "Updated %q", tlog.KObj{Obj: s.Current.Cluster})
+
+	// Wait until Cluster is updated in the cache.
+	// Note: We have to do this because otherwise using a cached client in the Reconcile func could
+	// return a stale state of the Cluster we just patched (because the cache might be stale).
+	// Note: It is good enough to check that the resource version changed. Other controllers might have updated the
+	// Cluster as well, but the combination of the patch call above without a conflict and a changed resource
+	// version here guarantees that we see the changes of our own update.
+	err = wait.PollUntilContextTimeout(ctx, 5*time.Millisecond, 5*time.Second, true, func(ctx context.Context) (bool, error) {
+		key := client.ObjectKey{Namespace: s.Current.Cluster.GetNamespace(), Name: s.Current.Cluster.GetName()}
+		cachedCluster := &clusterv1.Cluster{}
+		if err := r.Client.Get(ctx, key, cachedCluster); err != nil {
+			return false, err
+		}
+		return s.Current.Cluster.GetResourceVersion() != cachedCluster.GetResourceVersion(), nil
+	})
+	if err != nil {
+		return errors.Wrapf(err, "failed to patch %s: failed waiting for Cluster to be updated in cache", tlog.KObj{Obj: s.Current.Cluster})
```

Comment:
How is this "failed to patch" given we are beyond l439?
Also how does a wait here impact the controller workflow at all? As soon as the the patch request above or any other is sent a new even will trigger reconciliation and another controller worker thread will pick it, right?

Reply:
> How is this "failed to patch" given we are beyond l439?

Good point. You're right

> Also how does a wait here impact the controller workflow at all? As soon as the the patch request above or any other is sent a new event will trigger reconciliation and another worker thread of the controller will pick it, right?

Controller-runtime guarantees that there is at most one worker reconciling a specific object / cluster in this case. So while the Patch call will trigger a new reconcile (to be very specific by marking the Cluster item as dirty in the queue) this reconcile will wait until the current reconcile is finished